packages/data/src/components/with-dispatch/test/index.js

```diff
@@ -119,4 +119,50 @@ describe( 'withDispatch', () => {
 		expect( firstRegistryAction ).toHaveBeenCalledTimes( 2 );
 		expect( secondRegistryAction ).toHaveBeenCalledTimes( 2 );
 	} );
+
+	it( 'always calls select with the latest state in the handler passed to the component', () => {
+		const store = registry.registerStore( 'counter', {
+			reducer: ( state = 0, action ) => {
+				if ( action.type === 'update' ) {
+					return action.count;
+				}
+				return state;
+			},
+			actions: {
+				update: ( count ) => ( { type: 'update', count } ),
+			},
+			selectors: {
+				getCount: ( state ) => state,
+			},
+		} );
+
+		const Component = withDispatch( ( _dispatch, ownProps, _select ) => {
+			const outerCount = _select( 'counter' ).getCount();
+			return {
+				update: () => {
+					const innerCount = _select( 'counter' ).getCount();
+					expect( innerCount ).toBe( outerCount );
```

Comment:
I’m surprised that you don’t need to put selector inside handler to make it work properly. I anticipated that `outerCount` would store stale value.

Reply:
Actually, it might be correct. I think we use the original object when we call the handler to ensure `ownProps` are up to date. @aduth will know better :D